Voce hardening

arovinelli opened the pull request:
Implementation of  Voce Hardening law for the UO based CP framework.
close #11307.

Add a  new user object implementing the Voce hardening law.
Documentation and test are included

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
Voce Hardening Law for the crystal plasticity user object based framework

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
documentation included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

Commit:
close #11307

Implementation of the Voce hardening Law for the crystal plasticity UO based framework.
test and coumentation are included

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/178714/) on 14adbfc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11308/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11308/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 07bd65968c80797cc314fd7447433488dc3887b8`

andrsd commented:
```
$ git rebase --onto devel HEAD~4 voceHardening
$ git push -f <your-fork> voceHardening
```

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307

applied style patch

Commit:
close #11307
corrected an error in tests file

Commit:
close #11307
corrected an error in tests file

Commit:
close #11307
corrected an error in tests file

Commit:
close #11307
corrected an error in tests file

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/178735/) on 69036db wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11308/site)

This comment will be updated on new commits.

Commit:
close #11307
updated test input file to comply with documentation

permcody commented:
This now needs to be rebased. You still have several other commits in your new feature branch that are making this difficult to review.

arovinelli commented:
Ok, what do you want me exactly to do. When I rebase it following the procedure [http://mooseframework.org/wiki/Contributing/](url)  I got extra commits that do not belong to this branch.

permcody commented:
Checkout your branch (you are probably still siting on it). Make sure you have updated your remote tracking branches and rebase!

```
git checkout voceHardening
git fetch upstream
git rebase upstream/devel
```

Those last two commands are you get up to date with what's happening in the main repository. You'll want to run those frequently as you develop.

arovinelli commented:
Cody, i did what you say but things goes wrong. For some reason it gives me  a bunch of conflicts on my files. These conflicts are not supposed to be there as my local branch is what I pushed. If I solve the conflicts, it finds new ones after git rebase --continue. and the same is true for --skip.
Ideas are welcome.

arovinelli commented:
Ok, I've finally been able to rebase it. You migh want to take a look at it

permcody commented:
Great, I'll get to this soon!

dschwen commented:
Your equations in the docs are somehow broken:
http://mooseframework.inl.gov/docs/PRs/11308/site/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.html

Might be the extra empty lines.

dschwen commented on modules/tensor_mechanics/include/userobjects/CrystalPlasticityStateVarRateComponentVoce.h:
```diff
@@ -0,0 +1,72 @@
+  virtual bool calcStateVariableEvolutionRateComponent(unsigned int qp,
+                                                       std::vector<Real> & val) const;
+
+  // class containgn the avialbale slip systems
```
could you make those `///` comments so that doxygen picks those up. Also there are some typos in here :-)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/182532/) on 2ff1e9d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11308/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11308/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 870249bbc6bd43018e1cfd5746591015af40f27e`

arovinelli commented:
something went terribly wrong during rebase. let me fix this.
I will open a new branch push everything in there and delete this one.

Commit:
Update equations in the documentation. Updated comments in .h file
close #11307

Commit:
Update equations in the documentation. Updated comments in .h file
close #11307

permcody commented:
Ping @dschwen

dschwen commented on modules/tensor_mechanics/doc/content/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.md:
```diff
@@ -0,0 +1,54 @@
+This userobjects assumes that slip systems are provided in a well defined order and are grouped by planes (see  modules/tensor_mechanics/test/tests/cp_user_object/input_slip_sys_bcc48.txt for BCC crystals and modules/tensor_mechanics/test/tests/cp_user_object/input_slip_sys.txt for FCC)
+
+Seven variables need to be specified and one value is required for each group:
+\begin{itemize}
```
this does not render properly. Try using markdown here:

```
- item1
- more stuff
- and so on
```

renders as 

- item1
- more stuff
- and so on

arovinelli replied:
Done, hopefully it renders. By the way, how do I check offline that everything renders properly before pushing updates?

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,271 @@
+    _GroupGroup_Hardening_group_values(
+        getParam<std::vector<Real>>("GroupGroup_Hardening_group_values"))
+{
+  // perform input checks and initialize usefull variables
```
typo `useful`

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,271 @@
+        getParam<std::vector<Real>>("GroupGroup_Hardening_group_values"))
+{
+  // perform input checks and initialize usefull variables
+  _n_groups = _groups.size() - 1;
```
Why `- 1`? This is an unusual convention. We use `_n_xxx` variables a lot and they usually correspond directly to the `size()` of an underlying container.

dschwen replied:
By the way this is definitely wrong code. `_n_groups` is unsigned. So if `size()` is `0` it takes on a ginormous value (because it wraps around to the largest positive unsigned integer).

arovinelli replied:
Daniel, the variable _n_groups was set to be _groups.size() - 1 because in the input you definef group bounds. For example, let's say you are using BCC 24, and want to group slipsystems as primary and secondary then you will input 1 12 24. the format is 'format: [start end)'. You have 2 groups but 3 inputs. However I changed it to your standard notation. Now _n_groups = _groups.size().

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,271 @@
+CrystalPlasticityStateVarRateComponentVoce::checkHardeningParametersSize() const
+{
+  // check that at least one group exists
+  if (_n_groups <= 0)
```
`_n_groups` can never be < 0. It being 0 means only one group was specified. Is that invalid?

arovinelli replied:
Daniel, thanks a lot for spotting this mistake. I copy and pasted this from CrystalPlasticityStateVariable.C without double checking the code. Therefore this file also need to be checked. 

Anyway, Following the previous comment , this check is now
if (_n_groups < 2), because for one group to be defined we need at least two values.

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,271 @@
+  check_var_size &= _selfHardening_group_values.size() == _n_groups;
+  check_var_size &= _coplanarHardening_group_values.size() == _n_groups;
+  check_var_size &= _h0_group_values.size() == _n_groups;
+  check_var_size &= _GroupGroup_Hardening_group_values.size() == _n_groups * _n_groups;
```
I'd write this as

```
  const bool check_var_size =
      _h0_group_values.size() == _n_groups && _tau0_group_values.size() == _n_groups &&
      _tauSat_group_values.size() == _n_groups &&
      _hardeningExponent_group_values.size() == _n_groups &&
      _selfHardening_group_values.size() == _n_groups &&
      _coplanarHardening_group_values.size() == _n_groups && _h0_group_values.size() == _n_groups &&
      _GroupGroup_Hardening_group_values.size() == _n_groups * _n_groups;
```

but for a gold star you should really have a `paramError` for each parameter that fails the check. Otherwise the user will have to go through 8 input parameter vectors, count the elements in each one, and find the one that has the wron size (without any indication what the _right_ size is).

arovinelli replied:
Now each parameters has its own test.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/186593/) on 3937f4c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11308/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11308/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 84375a7ee7684e33b2e09e1c9d0f580d8c6963a6`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/186615/) on da64347 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11308/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11308/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 84375a7ee7684e33b2e09e1c9d0f580d8c6963a6`

Commit:
Updated documentation, now it should render properly.
Update the .C file such that each input parameter is tested separetly
and solved the issue with the _n_groups variable.

Commit:
Updated documentation, now it should render properly.
Update the .C file such that each input parameter is tested separetly
and solved the issue with the _n_groups variable.

Commit:
Applyed Patch

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/186641/) on 721ccfd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11308/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11308/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 84375a7ee7684e33b2e09e1c9d0f580d8c6963a6`

Commit:
Applyed Patch

Commit:
Applyed Patch

arovinelli commented:
I addressed all Daniel's comment. Please have a look at the new version.
All the best.

permcody commented:
You can look at the local site by following these instructions. Very useful while building docs. It auto-reloads as you save:
http://mooseframework.org/moose/utilities/MooseDocs/generate.html

arovinelli commented:
The "Non unity build" test failed, but it doesn't allow me to see the log.

permcody commented:
The non-unity failure is not your issue. Ignore it.

arovinelli committed:
Merge branch 'voceHardening' of https://github.com/arovinelli/moose into voceHardening

arovinelli commented:
I update the documentation and checked it my computer, now it renders properly.
Cody thanks for pointing me to that webpage to check the documentation offline. This is really useful!

permcody commented:
@dschwen - ready for review. SAM failure can be ignored.

dschwen commented on modules/tensor_mechanics/doc/content/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.md:
```diff
@@ -0,0 +1,59 @@
+
+   -  "selfHardening_group_values" $q^{\alpha\alpha}$
+
+   -  "coplanarHardening_group_values" $q^{\alpha\beta}$ for co-planar slip systems in the same group
```
This list does not render correctly: http://mooseframework.inl.gov/docs/PRs/11308/site/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.html#c69da0ed-3b16-4f50-b11f-19854863b059

Do not indent and do not leave empty lines between list items

dschwen commented on modules/tensor_mechanics/doc/content/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.md:
```diff
@@ -0,0 +1,59 @@
+
+## Description
+
+This UserObjects, `CrystalPlasticityStateVarRateComponentVoce` is meant to be used within the user object base crystal plasticity framework as a Voce hardening rule.
```
UserObjects -> UserObject

dschwen commented on modules/tensor_mechanics/doc/content/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.md:
```diff
@@ -0,0 +1,59 @@
+
+Like for the others crystal plasticity routines values of all the parameters can be specified by groups of slip systems.
+
+This userobjects assumes that slip systems are provided in a well defined order and are grouped by planes (see  modules/tensor_mechanics/test/tests/cp_user_object/input_slip_sys_bcc48.txt for BCC crystals and modules/tensor_mechanics/test/tests/cp_user_object/input_slip_sys.txt for FCC)
```
userobjects -> userobject

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,288 @@
+  params.addClassDescription(
+      "Phenomenological Voce constitutive model state"
+      " variable evolution rate "
+      "component base class.  Override this virtual functions in your class");
```
I don't think the last sentence makes sense in this description.

dschwen commented on modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C:
```diff
@@ -0,0 +1,288 @@
+CrystalPlasticityStateVarRateComponentVoce::initSlipSystem_GroupID(
+    std::vector<unsigned int> & _slipSystem_GroupID) const
+// this routine is generate a vector containing the association between
+// slip system number and provided group edges
```
this comment seems out of place. I suggest putting it into the header file where the member is declared. That way it gets picked up by the doxugen documentation system (make sure you use `///` triple-slash comment)

dschwen commented on modules/tensor_mechanics/doc/content/documentation/systems/UserObjects/CrystalPlasticityStateVarRateComponentVoce.md:
```diff
@@ -0,0 +1,59 @@
+
+   -  "coplanarHardening_group_values" $q^{\alpha\beta}$ for co-planar slip systems in the same group
+
+   -  "GroupGroup_Hardening_group_values" $q^{\alpha\beta}$ This parameter requires NxN values with N being the number of groups. Values are listed as $value_ij$ (e.g. ij=11,12,21,22) with $i$ being the actual group and $j$ the secondary group. For $i==j$ the value represents the latent hardening coefficient between one slip system and all the non co-planar ones belonging the the same group. For  ...
```
There are better latex ways to express `==` AND `!==`, use `=` and `\neq`

dschwen reviewed (commented):
Some comments about typos and documentation.

dschwen commented:
By the way this needs to be squashed (can be done during merging)

dschwen commented:
Actually, @arovinelli , if you want I can merge this as is and just do the small fixes myself. You've already spent a ton on time on this :-)
